### PR TITLE
support type modifiers in ORCA

### PR DESCRIPTION
**Pull request by d**
Sister PR at https://github.com/greenplum-db/gpdb/pull/4510

### CODE REVIEW

File: server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp [hunk 1]
```diff
@@ -275,7 +275,7 @@ namespace
 
 			void AddIntColumn(CWStringConst strColumnName, int iAttno, BOOL fNullable)
 			{
-				CColumnDescriptor *pcoldesc = GPOS_NEW(m_pmp) CColumnDescriptor(m_pmp, m_pmdtypeint4, CName(&strColumnName), iAttno, fNullable);
+				CColumnDescriptor *pcoldesc = GPOS_NEW(m_pmp) CColumnDescriptor(m_pmp, m_pmdtypeint4, -1 /* iTypeModifier */, CName(&strColumnName), iAttno, fNullable);
```
Comment (vraghavan78) [inline server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp]: can we have this as a macro?

Comment (hardikar) [reply server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp]: Check again. PR was opened a little too early..

File: libgpopt/include/gpopt/operators/CScalarFunc.h [hunk 2]
```diff
@@ -43,6 +43,8 @@ namespace gpopt
 			// return type
 			IMDId *m_pmdidRetType;
 
+			const INT m_iRetTypeModifier;
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/operators/CScalarFunc.h]: for consistency add a comment, else remove all useless comments

File: libgpopt/src/base/CUtils.cpp [hunk 3]
```diff
@@ -1972,24 +1972,6 @@ CUtils::PexprAggFunc
 	return GPOS_NEW(pmp) CExpression(pmp, popScAggFunc, pdrgpexpr);
 }
 
-// construct a scalar function
-CExpression *
-CUtils::PexprScalarFunc
```
Comment (vraghavan78) [inline libgpopt/src/base/CUtils.cpp]: you don't need this any more?

File: libnaucrates/include/naucrates/dxl/operators/CDXLDatum.h [hunk 4]
```diff
@@ -93,10 +96,16 @@ namespace gpdxl
 				return m_pmdidType;
 			}
 
+			INT
+			ITypeModifier() const
+			{
+				return m_iTypeModifier;
+			}
+
```
Comment (bhuvnesh2703) [inline libnaucrates/include/naucrates/dxl/operators/CDXLDatum.h]: all the other methods have the implementations in cpp, so why not move it there.

File: libgpopt/src/base/CColumnFactory.cpp [hunk 5]
```diff
@@ -193,18 +196,14 @@ CColumnFactory::PcrCreate
 	ULONG ulOpSource
 	)
 {
-	CName *pnameCopy = GPOS_NEW(m_pmp) CName(m_pmp, name); 
-	CAutoP<CName> a_pnameCopy(pnameCopy);
-
```
Comment (d) [inline libgpopt/src/base/CColumnFactory.cpp]: Duh, I just realized why the `CAutoP` was needed here: if the `GPOS_NEW` below throws, we'd be leaking the `CName` object here.

The pattern of using `CAutoP<T>::PtReset()` is really indicative of needing something of a move semantic in our codebase, but that's out of scope of this PR...

**Pull request merged**